lib/flask_tasks.py

```diff
@@ -0,0 +1,194 @@
+        with current_app.test_request_context():
+            return _url_for(*args, **kwargs)
+    return _url_for(*args, **kwargs)
+
+
+@tasks_bp.before_app_first_request
+def before_first_request():
+    """Start a background thread that cleans up old tasks."""
+    def clean_old_tasks():
+        logging.debug("Cleanup old tasks")
+        """
+        This function cleans up old tasks from our in-memory data structure.
+        """
+        global tasks
+        while True:
+            # Only keep tasks that are running or that finished less than 5
+            # minutes ago.
+            five_min_ago = timestamp() - 5 * 60
+            tasks = {id: task for id, task in tasks.items()
+                     if 't' not in task or task['t'] > five_min_ago}
+            time.sleep(60)
+
+    if not testing:
+        logging.debug("Production mode, run cleanups")
+        thread = threading.Thread(target=clean_old_tasks)
+        #thread.start()
+    else:
+        logging.debug("Testing mode, no cleanup!")
+
+
+def async_task(f):
+    """
+    This decorator transforms a sync route to asynchronous by running it
+    in a background thread.
+    """
+    @wraps(f)
+    def wrapped(*args, **kwargs):
+        def task(app, environ):
+            # Create a request context similar to that of the original request
+            # so that the task can have access to flask.g, flask.request, etc.
+            with app.request_context(environ):
+                try:
+                    # Run the route function and record the response
+                    logging.debug("____________________________")
+                    logging.debug("Executing target function")
+                    tasks[id]['started'] = True
+                    tasks[id]['rv'] = f(*args, **kwargs)
+                    tasks[id]['status'] = 'done'
+                    logging.debug("target function completed")
+                    logging.debug("____________________________")
+                except HTTPException as e:
+                    logging.debug("Caught http exception:" + str(e))
+                    tasks[id]['rv'] = current_app.handle_http_exception(e)
+                    traceback.print_exc(file=sys.stdout)
+                except Exception as e:
+                    logging.debug("Caught exception:" + str(e))
+                    # The function raised an exception, so we set a 500 error
+                    tasks[id]['rv'] = InternalServerError()
+                    traceback.print_exc(file=sys.stdout)
+
+            #                    if current_app.debug:
+#                        # We want to find out if something happened so reraise
+#                        raise
+                finally:
+                    # We record the time of the response, to help in garbage
+                    # collecting old tasks
+                    tasks[id]['endtime'] = timestamp()
+
+        # Assign an id to the asynchronous task
+        id = uuid.uuid4().hex
```

Comment:
this looks like ambiguous usage of id() python function, maybe it is better to use 'task_id' variable name?

Reply:
done